app/services/fact_parser.rb

```diff
@@ -33,18 +33,38 @@ def operatingsystem
   end
 
   def environment
+    name = environment_name
+    Environment.unscoped.where(:name => name).first_or_create
+  end
+
+  def environment_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def architecture
+    name = architecture_name
+    Architecture.where(:name => name).first_or_create if name.present?
+  end
+
+  def architecture_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def model
+    name = model_name
+    Model.where(:name => name.strip).first_or_create if name.present?
+  end
+
+  def model_name
     raise NotImplementedError, not_implemented_error(__method__)
   end
 
   def domain
+    name = domain_name
+    Domain.unscoped.where(:name => name).first_or_create if name.present?
```

Comment:
Prior discussion: https://github.com/theforeman/foreman/pull/8449#discussion_r614895291

tl;dr: unscoped is a risk, but it was in the Puppet fact parser where I copied it from. [foreman_ansible](https://github.com/theforeman/foreman_ansible/blob/5eec05f645eae1b1cde9de2179f4cf5208873ae3/app/services/foreman_ansible/fact_parser.rb#L28-L32) doesn't use unscoped. There may be more out there.